### PR TITLE
Fix build in nix shell

### PR DESCRIPTION
**Pull request by Ericson2314**
Includes https://github.com/NixOS/nix/pull/3165, which brings in new master commits.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -476,9 +476,12 @@ endif
...
+if (not get_option('with_libsodium').disabled())
+    libsodium_dep_raw = cpp.find_library('sodium', required: get_option('with_libsodium').enabled())
+endif
+if (not get_option('with_libsodium').disabled() and libsodium_dep_raw.found())
     libsodium_dep = declare_dependency(
-        dependencies : cpp.find_library('sodium'),
+        dependencies : libsodium_dep_raw,
         link_args :  get_option('sodium_link_args'))
     config_h.set('HAVE_SODIUM', 1, description : 'Whether to use libsodium for cryptography.')
 else
```
Comment (p01arst0rm) [inline meson.build]: this is really bad for readability, surely theres a nicer/simpler way to write this

File: meson.build [hunk 2]
```diff
@@ -163,7 +163,8 @@ if (cpp.has_header('sys/ndir.h'))
         description : 'Define to 1 if you have the <sys/ndir.h> header file, and it defines `DIR`')
 endif
 
-if (cpp.has_header('dirent.h'))
+has_dirent_h = cpp.has_header('dirent.h')
+if (has_dirent_h)
```
Comment (p01arst0rm) [inline meson.build]: nice spot ^^

File: meson.build [hunk 3]
```diff
@@ -309,6 +310,23 @@ if meson.get_compiler('cpp').compiles(
...
+  #include <sys/types.h>
+  #include <dirent.h>
+'''
+
+# Code taken from glib, thanks!
+if has_dirent_h and meson.get_compiler('cpp').has_member('struct dirent', 'd_type', prefix: dirent_h_prefix)
+  define = 'HAVE_STRUCT_@0@_@1@'.format('dirent'.to_upper(), 'd_type'.underscorify().to_upper())
+  config_h.set(define, 1)
+endif
+
```
Comment (p01arst0rm) [inline meson.build]: this doesnt quite follow the code style for the rest of the build files, and it doesnt explain whats happening here. theres also no description for the config header

File: meson.build [hunk 4]
```diff
@@ -330,8 +348,8 @@ tar = find_program('tar', required : true)
 bzip2 = find_program('bzip2', required : true)
 gzip = find_program('gzip', required : true)
 xz = find_program('xz', required : true)
-dot = find_program('dot', required : true)
-lsof = find_program('lsof', required : true)
+dot = find_program('dot', required : false)
+lsof = find_program('lsof', required : false)
```
Comment (p01arst0rm) [inline meson.build:350]: lsof is required by libnixstore iirc..?

File: meson_options.txt [hunk 5]
```diff
@@ -153,8 +153,8 @@ option(
 
 option(
     'with_libsodium',
-    type : 'boolean',
-    value : 'true',
+    type : 'feature',
+    value : 'auto',
     description : 'build nix with shared libs')
```
Comment (p01arst0rm) [inline meson_options.txt]: was this description always wrong? :O

**Review by p01arst0rm — CHANGES_REQUESTED** [review]
theres some hackery, but good stuff ^^

File: meson.build (hunk 4 above)
Comment (Ericson2314) [reply meson.build:350]: So perhaps I didn't do this quite right, but it just needs it an runtime, and will do a PATH lookup then if it doesn't find it. This what the old version did.

File: meson_options.txt (hunk 5 above)
Comment (Ericson2314) [reply meson_options.txt]: Oh haha yeah it was.

File: meson.build (hunk 3 above)
Comment (Ericson2314) [reply meson.build]: So I think the header checks should go *after* the dependency checks, so the headers checks and use the dependencies (good for weird dirs, etc.)

I didn't want to put that much code motion in this PR, but maybe you rather I do for stylistic consistency?

Comment (p01arst0rm) [reply meson.build]: mmm good point, maybe re order things in a housecleaning PR later

yep! the high verbosity of the files was purposeful, and id like to see it maintained into the future as much as possible

File: meson_options.txt (hunk 5 above)
Comment (Ericson2314) [reply meson_options.txt]: @p01arst0rm ah did you mean for me to fix this too, now that we are noticing it?

Comment (p01arst0rm) [reply meson_options.txt]: yep ^^

**Comment by Ericson2314** [comment]
@p01arst0rm OK good to go now?

File: meson.build (hunk 3 above)
Comment (p01arst0rm) [reply meson.build]: This is still not very clear, it should really be restructured.

File: meson.build [hunk 6]
```diff
@@ -361,13 +379,13 @@ description : 'Whether link() works on symlinks')
 #--------------------------------------------------
 boost_dep = declare_dependency(
     dependencies : [
-        cpp.find_library('libboost_system', dirs : libdir),
-        cpp.find_library('libboost_context', dirs : libdir),
-        cpp.find_library('libboost_thread', dirs : libdir)],
+        cpp.find_library('boost_system'),
+        cpp.find_library('boost_context'),
+        cpp.find_library('boost_thread')],
```
Comment (p01arst0rm) [inline meson.build:382]: This is actually a regression for FreeBSD building. Did you check if it compiles on FreeBSD?

File: meson.build [hunk 7]
```diff
@@ -390,7 +408,7 @@ libbrotli_dep = declare_dependency(
 # Look for OpenSSL, a required dependency.
 #--------------------------------------------------
 openssl_dep = declare_dependency(
-    dependencies: cpp.find_library('libssl', dirs : libdir),
+    dependencies: cpp.find_library('ssl'),
```
Comment (p01arst0rm) [inline meson.build:409]: Another regression.

File: meson.build [hunk 8]
```diff
@@ -418,14 +436,14 @@ pthread_dep = declare_dependency(
 # Look for libdl, a required dependency.
 #--------------------------------------------------
 libdl_dep = declare_dependency(
-    dependencies : cpp.find_library('dl', dirs : libdir),
+    dependencies : cpp.find_library('dl'),
```
Comment (p01arst0rm) [inline meson.build:437]: another regression

File: meson.build [hunk 9]
```diff
@@ -418,14 +436,14 @@ pthread_dep = declare_dependency(
...
     link_args :  get_option('dl_link_args'))
 
 
 # Look for libbz2, a required dependency.
 #--------------------------------------------------
 libbz2_dep = declare_dependency(
-    dependencies : cpp.find_library('bz2', dirs : libdir),
+    dependencies : cpp.find_library('bz2'),
```
Comment (p01arst0rm) [inline meson.build:444]: :)

File: meson.build [hunk 10]
```diff
@@ -437,7 +455,7 @@ libbz2_dep = declare_dependency(
 # editline.h when the pkg-config approach fails.
 
 editline_dep = declare_dependency(
-    dependencies : cpp.find_library('libeditline'),
+    dependencies : cpp.find_library('editline'),
```
Comment (p01arst0rm) [inline meson.build:456]: this breaks building on arch too

File: meson.build [hunk 11]
```diff
@@ -458,13 +476,14 @@ endif
...
+if (libsodium_dep_raw.found())
     libsodium_dep = declare_dependency(
-        dependencies : cpp.find_library('libsodium', dirs : libdir),
+        dependencies : libsodium_dep_raw,
         link_args :  get_option('sodium_link_args'))
     config_h.set('HAVE_SODIUM', 1, description : 'Whether to use libsodium for cryptography.')
 else
-    libsodium_dep = declare_dependency()
+    libsodium_dep = dependency('', required: false)
 endif
```
Comment (p01arst0rm) [inline meson.build]: This is an overcomplication and makes the whole check messier (imo) with no real benefit

**Review by p01arst0rm — CHANGES_REQUESTED** [review]

File: meson.build (hunk 11 above)
Comment (Ericson2314) [reply meson.build]: Which part is?

Comment (p01arst0rm) [reply meson.build]: line 479,480,482. there is no need to declare the dependency twice in this way just to encompass the required inside the dependency. required defaults to true.

File: meson.build (hunk 3 above)
Comment (p01arst0rm) [reply meson.build]: better :)

**Comment by p01arst0rm** [comment]
LGTM :)

**Pull request merged**